src/TypeInference/Coalescing.hs

```diff
@@ -96,56 +97,66 @@ coalesce result@MkSolverResult { tvarSolution, kvarSolution } = MkBisubstitution
 coalesceType :: Typ pol -> CoalesceM (Typ pol)
 coalesceType (TySkolemVar loc rep mk tv) =  do
   return (TySkolemVar loc rep mk tv)
-coalesceType (TyRecVar loc rep mk tv) = do 
-  return (TyRecVar loc rep mk tv)
+coalesceType (TyRecVar loc rep pk tv) = do 
+  return (TyRecVar loc rep pk tv)
 coalesceType (TyUniVar _ PosRep mk tv) = do
-    isInProcess <- inProcess (tv, Pos)
-    if isInProcess
-        then do
-            recVar <- getOrElseUpdateRecVar (tv, Pos)
-            return (TyRecVar defaultLoc PosRep mk recVar)
-        else do
-            VariableState { vst_lowerbounds } <- getVariableState tv
-            let f r = r { r_inProcess =  S.insert (tv, Pos) (r_inProcess r) }
-            lbs' <- local f $ sequence $ coalesceType <$> vst_lowerbounds
-            recVarMap <- gets s_recursive
-            case M.lookup (tv, Pos) recVarMap of
-                Nothing     -> do
-                    newName <- getSkolemVar tv
-                    return $                                            mkUnion defaultLoc mk (TySkolemVar defaultLoc PosRep mk newName : lbs')
-                Just recVar -> return $ TyRec defaultLoc PosRep recVar (mkUnion defaultLoc mk (TyRecVar defaultLoc PosRep mk recVar  : lbs'))
+  isInProcess <- inProcess (tv, Pos)
+  if isInProcess then
+    case mk of 
+      CBox eo -> do
+        recVar <- getOrElseUpdateRecVar (tv, Pos)
+        return (TyRecVar defaultLoc PosRep (MkPolyKind [] eo) recVar)
```

Comment:
Here I don't have access to the arguments of the polykind for the recursive variable (happens the same for the other occurences here. 
To fix this, Univars might require a polykind as well, but since univars can also have primitive kinds, this would require a lot of changes  and should probably happen in a different PR